java/src/main/scala/com/spotify/featran/java/JavaOps.scala

```diff
@@ -65,4 +65,71 @@ private object JavaOps {
   def featureValuesDouble[T](fe: FeatureExtractor[JList, T]): JList[Array[Double]] =
     fe.featureValues[Array[Double]]
 
+  import JavaFeatureBuilder._
+  def featureValuesFloatSparseArray[T](fe: FeatureExtractor[JList, T])
+  : JList[FloatSparseArray] = fe.featureValues[FloatSparseArray]
+  def featureValuesDoubleSparseArray[T](fe: FeatureExtractor[JList, T])
+  : JList[DoubleSparseArray] = fe.featureValues[DoubleSparseArray]
+
+}
+
+case class FloatSparseArray(indices: Array[Int], values: Array[Float], length: Int)
+case class DoubleSparseArray(indices: Array[Int], values: Array[Double], length: Int)
+
+/*private*/ object JavaFeatureBuilder {
+
+  case class SparseArray[@specialized (Float, Double) T]
+  (indices: Array[Int], values: Array[T], length: Int)
+
+  implicit val floatSparseArrayFB: FeatureBuilder[FloatSparseArray] =
+    sparseArrayFB[Float].map(a => FloatSparseArray(a.indices, a.values, a.length))
+  implicit val doubleSparseArrayFB: FeatureBuilder[DoubleSparseArray] =
+    sparseArrayFB[Double].map(a => DoubleSparseArray(a.indices, a.values, a.length))
+
+  def sparseArrayFB[@specialized (Float, Double) T: ClassTag : FloatingPoint]
+  : FeatureBuilder[SparseArray[T]] = new FeatureBuilder[SparseArray[T]] {
+    private val initCapacity = 1024
+    private var dim: Int = _
+    private var offset: Int = 0
+    private var i: Int = 0
+    private var indices: Array[Int] = _
+    private var values: Array[T] = _
+    private val fp = implicitly[FloatingPoint[T]]
+    override def init(dimension: Int): Unit = {
+      dim = dimension
+      offset = 0
+      i = 0
+      if (indices == null) {
+        val n = math.min(dimension, initCapacity)
+        indices = new Array[Int](n)
+        values = new Array[T](n)
+      }
+    }
+    override def add(name: String, value: Double): Unit = {
+      indices(i) = offset
+      values(i) = fp.fromDouble(value)
+      i += 1
+      offset += 1
+      if (indices.length == i) {
```

Comment:
Any reason not to use ArrayBuffer?  I believe the behavior you have coded here duplicates that class.   You can also give it a hintSize with the n from above.

Reply:
Duh I'll look into it.

Reply:
Actually no, since `ArrayBuffer` doesn't specialize for primitive and will use boxed `Integer`, `Float`, `Double`, etc., and I'll have to duplicate code to manually specialize them.